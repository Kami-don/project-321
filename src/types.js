// @flow
import type { BoxModel, Rect, Position } from 'css-box-model';

export type Id = string;
export type DraggableId = Id;
export type DroppableId = Id;
export type TypeId = Id;
export type ZIndex = number | string;

export type DroppableDescriptor = {|
  id: DroppableId,
  type: TypeId,
|};

export type DraggableDescriptor = {|
  id: DraggableId,
  index: number,
  // Inherited from Droppable
  droppableId: DroppableId,
  // This is technically redundant but it avoids
  // needing to look up a parent droppable just to get its type
  type: TypeId,
|};

export type Direction = 'horizontal' | 'vertical';

export type VerticalAxis = {|
  direction: 'vertical',
  line: 'y',
  start: 'top',
  end: 'bottom',
  size: 'height',
  crossAxisLine: 'x',
  crossAxisStart: 'left',
  crossAxisEnd: 'right',
  crossAxisSize: 'width',
|};

export type HorizontalAxis = {|
  direction: 'horizontal',
  line: 'x',
  start: 'left',
  end: 'right',
  size: 'width',
  crossAxisLine: 'y',
  crossAxisStart: 'top',
  crossAxisEnd: 'bottom',
  crossAxisSize: 'height',
|};

export type Axis = VerticalAxis | HorizontalAxis;

export type Placeholder = {|
  client: BoxModel,
  tagName: string,
  display: string,
|};

export type DraggableDimension = {|
  descriptor: DraggableDescriptor,
  // the placeholder for the draggable
  placeholder: Placeholder,
  // relative to the viewport when the drag started
  client: BoxModel,
  // relative to the whole page
  page: BoxModel,
|};

export type Scrollable = {|
  // This is the window through which the droppable is observed
  // It does not change during a drag
  framePageMarginBox: Rect,
  // Whether or not we should clip the subject by the frame
  // Is controlled by the ignoreContainerClipping prop
  shouldClipSubject: boolean,
  scroll: {|
    initial: Position,
    current: Position,
    // the maximum allowable scroll for the frame
    max: Position,
    diff: {|
      value: Position,
      // The actual displacement as a result of a scroll is in the opposite
      // direction to the scroll itself. When scrolling down items are displaced
      // upwards. This value is the negated version of the 'value'
      displacement: Position,
    |},
  |},
|};

export type DroppableDimensionViewport = {|
  // will be null if there is no closest scrollable
  closestScrollable: ?Scrollable,
  subjectPageMarginBox: Rect,
  // this is the subject through the viewport of the frame (if applicable)
  // it also takes into account any changes to the viewport scroll
  // clipped area will be null if it is completely outside of the frame and frame clipping is on
  clippedPageMarginBox: ?Rect,
|};

export type DroppableDimension = {|
  descriptor: DroppableDescriptor,
  axis: Axis,
  isEnabled: boolean,
  // relative to the current viewport
  client: BoxModel,
  // relative to the whole page
  page: BoxModel,
  // The container of the droppable
  viewport: DroppableDimensionViewport,
|};
export type DraggableLocation = {|
  droppableId: DroppableId,
  index: number,
|};

export type DraggableDimensionMap = { [key: DraggableId]: DraggableDimension };
export type DroppableDimensionMap = { [key: DroppableId]: DroppableDimension };

export type Displacement = {|
  draggableId: DraggableId,
  isVisible: boolean,
  shouldAnimate: boolean,
|};

export type DragMovement = {|
  // The draggables that need to move in response to a drag.
  // Ordered by closest draggable to the *current* location of the dragging item
  displaced: Displacement[],
  amount: Position,
  // is moving forward relative to the starting position
  // TODO: rename to 'shouldDisplaceForward'?
  isBeyondStartPosition: boolean,
|};

export type DragImpact = {|
  movement: DragMovement,
  // the direction of the Droppable you are over
  direction: ?Direction,
  destination: ?DraggableLocation,
|};

export type ItemPositions = {|
  // where the user initially selected
  // This point is not used to calculate the impact of a dragging item
  // It is used to calculate the offset from the initial selection point
  selection: Position,
  // the current center of the item
  borderBoxCenter: Position,
  // how far the item has moved from its original position
  offset: Position,
|};

// When dragging with a pointer such as a mouse or touch input we want to automatically
// scroll user the under input when we get near the bottom of a Droppable or the window.
// When Dragging with a keyboard we want to jump as required
export type AutoScrollMode = 'FLUID' | 'JUMP';

// export type Viewport = {|
//   scroll: Position,
//   maxScroll: Position,
//   subject: Rect,
// |}

export type DragPositions = {|
  client: ItemPositions,
  page: ItemPositions,
|};

// published when a drag starts
export type DragStart = {|
  draggableId: DraggableId,
  type: TypeId,
  source: DraggableLocation,
|};

export type DragUpdate = {|
  ...DragStart,
  // may not have any destination (drag to nowhere)
  destination: ?DraggableLocation,
|};

export type DropReason = 'DROP' | 'CANCEL';

// published when a drag finishes
export type DropResult = {|
  ...DragUpdate,
  reason: DropReason,
|};

export type PendingDrop = {|
  // TODO: newHomeClientOffset
  newHomeOffset: Position,
  impact: DragImpact,
  result: DropResult,
|};

export type ScrollOptions = {|
  shouldPublishImmediately: boolean,
|};

// using the draggable id rather than the descriptor as the descriptor
// may change as a result of the initial flush. This means that the lift
// descriptor may not be the same as the actual descriptor. To avoid
// confusion the request is just an id which is looked up
// in the dimension-marshal post-flush
// Not including droppableId as it might change in a drop flush
export type LiftRequest = {|
  draggableId: DraggableId,
  scrollOptions: ScrollOptions,
|};

export type Critical = {|
  draggable: DraggableDescriptor,
  droppable: DroppableDescriptor,
|};

export type Viewport = {|
  // live updates with the latest values
  frame: Rect,
  scroll: {|
    initial: Position,
    current: Position,
    max: Position,
    diff: {|
      value: Position,
      // The actual displacement as a result of a scroll is in the opposite
      // direction to the scroll itself.
      displacement: Position,
    |},
  |},
|};

export type DimensionMap = {|
  draggables: DraggableDimensionMap,
  droppables: DroppableDimensionMap,
|};

export type Publish = {|
  additions: {|
    draggables: DraggableDimension[],
    droppables: DroppableDimension[],
  |},
  // additions: DimensionMap,
  removals: {|
    draggables: DraggableId[],
    droppables: DroppableId[],
  |},
|};

export type IdleState = {|
  phase: 'IDLE',
|};

export type PreparingState = {|
  phase: 'PREPARING',
|};

export type DraggingState = {|
  phase: 'DRAGGING',
  isDragging: true,
  critical: Critical,
  autoScrollMode: AutoScrollMode,
  dimensions: DimensionMap,
  initial: DragPositions,
  current: DragPositions,
  impact: DragImpact,
  viewport: Viewport,
  // if we need to jump the scroll (keyboard dragging)
  scrollJumpRequest: ?Position,
  // whether or not draggable movements should be animated
  shouldAnimate: boolean,
|};

// While dragging we can enter into a bulk collection phase
// During this phase no drag updates are permitted.
// If a drop occurs during this phase, it must wait until it is
// completed before continuing with the drop
// TODO: rename to BulkCollectingState
export type CollectingState = {|
  ...DraggingState,
  phase: 'COLLECTING',
|};

// If a drop action occurs during a bulk collection we need to
// wait for the collection to finish before performing the drop.
// This is to ensure that everything has the correct index after
// a drop
export type DropPendingState = {|
  ...DraggingState,
  phase: 'DROP_PENDING',
  isWaiting: boolean,
  reason: DropReason,
|};

// An optional phase for animating the drop / cancel if it is needed
export type DropAnimatingState = {|
  phase: 'DROP_ANIMATING',
  pending: PendingDrop,
  // We still need to render placeholders and fix the dimensions of the dragging item
  dimensions: DimensionMap,
|};

export type State =
  | IdleState
  | PreparingState
  | DraggingState
  | CollectingState
  | DropPendingState
  | DropAnimatingState;

export type Announce = (message: string) => void;

export type HookProvided = {|
  announce: Announce,
|};

export type OnDragStartHook = (
  start: DragStart,
  provided: HookProvided,
) => void;
export type OnDragUpdateHook = (
  update: DragUpdate,
  provided: HookProvided,
) => void;
export type OnDragEndHook = (
  result: DropResult,
  provided: HookProvided,
) => void;

export type Hooks = {|
  onDragStart?: OnDragStartHook,
  onDragUpdate?: OnDragUpdateHook,
  // always required
  onDragEnd: OnDragEndHook,
|};
