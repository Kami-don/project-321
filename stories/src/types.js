// @flow
import type { DraggableId, DraggableLocation } from '../../src/types';

export type Author = {|
  id: string,
  name: string,
  avatarUrl: string,
  url: string,
|}

export type Quote = {|
  id: string,
  content: string,
  author: Author
|}

export type Dragging = {|
  id: DraggableId,
  location: DraggableLocation,
|}

export type QuoteMap = {
  [key: string]: Quote[]
}
