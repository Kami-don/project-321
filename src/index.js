// @flow

// Components

export { default as DragDropContext } from './view/drag-drop-context';
export { default as Droppable } from './view/droppable';
export { default as Draggable } from './view/draggable';

// Utils

export { resetServerContext } from './view/drag-drop-context';

// Public flow types

export type {
  Id,
  TypeId,
  DraggableId,
  DroppableId,
  // Hooks
  DragStart,
  DragUpdate,
  DropResult,
  HookProvided,
  Announce,
  DraggableLocation,
  OnDragStartHook,
  OnDragUpdateHook,
  OnDragEndHook,
} from './types';

// Droppable
export type {
  Provided as DroppableProvided,
  StateSnapshot as DroppableStateSnapshot,
  DroppableProps,
} from './view/droppable/droppable-types';

// Draggable
export type {
  Provided as DraggableProvided,
  StateSnapshot as DraggableStateSnapshot,
  DraggableProps,
  DraggableStyle,
  DraggingStyle,
  NotDraggingStyle,
} from './view/draggable/draggable-types';

// DragHandle
export type { DragHandleProps } from './view/drag-handle/drag-handle-types';
