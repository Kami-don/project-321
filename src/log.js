// @flow
/* eslint-disable no-console */
export function log(...args: mixed[]) {
  if (process.env.NODE_ENV !== 'production') {
    console.log(...args);
  }
}

export function warn(...args: mixed[]) {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(...args);
  }
}

export function error(...args: mixed[]) {
  if (process.env.NODE_ENV !== 'production') {
    console.error(...args);
  }
}

export function group(title: string = '') {
  if (process.env.NODE_ENV !== 'production') {
    console.group(title);
  }
}

export function groupEnd() {
  if (process.env.NODE_ENV !== 'production') {
    console.groupEnd();
  }
}
