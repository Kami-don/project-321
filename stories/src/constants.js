// @flow

export const grid: number = 8;
export const borderRadius: number = 2;
