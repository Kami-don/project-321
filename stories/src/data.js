// @flow
import type { Author, Quote, QuoteMap } from './types';

const jake: Author = {
  id: '1',
  name: 'Jake',
  url: 'http://adventuretime.wikia.com/wiki/Jake',
  avatarUrl: 'https://68.media.tumblr.com/avatar_1f7bdbbeb59c_128.png',
};

const BMO: Author = {
  id: '2',
  name: 'BMO',
  url: 'http://adventuretime.wikia.com/wiki/BMO',
  avatarUrl: 'https://68.media.tumblr.com/avatar_1a34fe6de498_128.png',
};

const finn: Author = {
  id: '3',
  name: 'Finn',
  url: 'http://adventuretime.wikia.com/wiki/Finn',
  avatarUrl: 'https://68.media.tumblr.com/avatar_09404f3287c6_128.png',
};

const princess: Author = {
  id: '4',
  name: 'Princess bubblegum',
  url: 'http://adventuretime.wikia.com/wiki/Princess_Bubblegum',
  avatarUrl: 'https://68.media.tumblr.com/avatar_ec98529441c4_128.png',
};

export const authors: Author[] = [
  jake, BMO, finn, princess,
];

export const quotes: Quote[] = [
  {
    id: '1',
    content: 'Sometimes life is scary and dark',
    author: BMO,
  },
  {
    id: '2',
    content: 'Sucking at something is the first step towards being sorta good at something.',
    author: jake,
  },
  {
    id: '3',
    content: 'You got to focus on what\'s real, man',
    author: jake,
  },
  {
    id: '4',
    content: 'Is that where creativity comes from? From sad biz?',
    author: finn,
  },
  {
    id: '5',
    content: 'Homies help homies. Always',
    author: finn,
  },
  {
    id: '6',
    content: 'Responsibility demands sacrifice',
    author: princess,
  },
  {
    id: '7',
    content: 'That\'s it! The answer was so simple, I was too smart to see it!, That\'s it! The answer was so simple, I was too smart to see it!, That\'s it! The answer was so simple, I was too smart to see it!, That\'s it! The answer was so simple, I was too smart to see it!',
    author: princess,
  },
  {
    id: '8',
    content: 'People make mistakes. Itâ€™s a part of growing up',
    author: finn,
  },
  {
    id: '9',
    content: 'Don\'t you always call sweatpants \'give up on life pants,\' Jake?',
    author: finn,
  },
  {
    id: '10',
    content: 'I should not have drunk that much tea!',
    author: princess,
  },
  {
    id: '11',
    content: 'Please! I need the real you!',
    author: princess,
  },
  {
    id: '12',
    content: 'Haven\'t slept for a solid 83 hours, but, yeah, I\'m good.',
    author: princess,
  },
];

let idCount: number = 0;

export const getQuotes = (count: number): Quote[] =>
  Array.from({ length: count }, (v, k) => k).map(() => {
    const random: Quote = quotes[Math.floor(Math.random() * quotes.length)];

    const custom: Quote = {
      id: `quote-${idCount++}`,
      content: random.content,
      author: random.author,
    };

    return custom;
  });

export const getAuthors = (count: number): Author[] =>
  Array.from({ length: count }, (v, k) => k).map(() => {
    const random: Author = authors[Math.floor(Math.random() * authors.length)];

    const custom: Author = {
      id: `author-${idCount++}`,
      name: random.name,
      avatarUrl: random.avatarUrl,
      url: random.url,
    };

    return custom;
  });

const getByAuthor = (author: Author, items: Quote[]): Quote[] =>
  items.filter((quote: Quote) => quote.author === author);

export const authorQuoteMap: QuoteMap =
  authors.reduce((previous: QuoteMap, author: Author) => ({
    ...previous,
    [author.name]: getByAuthor(author, quotes),
  }), {});

export const generateQuoteMap = (total: number): QuoteMap => authors.reduce(
  (previous: QuoteMap, author: Author) => ({
    ...previous,
    [author.name]: getQuotes(total / authors.length),
  }),
  {}
);
