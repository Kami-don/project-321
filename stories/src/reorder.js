// @flow
import type { Quote, QuoteMap } from './types';
import type { DraggableLocation } from '../../src/types';

// a little function to help us with reordering the result
const reorder = (
  list: any[],
  startIndex: number,
  endIndex: number): any[] => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};

export default reorder;

type ReorderQuoteMapArgs = {|
  quoteMap: QuoteMap,
  source: DraggableLocation,
  destination: DraggableLocation,
|}

export type ReorderQuoteMapResult = {|
  quoteMap: QuoteMap,
|}

export const reorderQuoteMap = ({
  quoteMap,
  source,
  destination,
}: ReorderQuoteMapArgs): ReorderQuoteMapResult => {
  const current: Quote[] = [...quoteMap[source.droppableId]];
  const next: Quote[] = [...quoteMap[destination.droppableId]];
  const target: Quote = current[source.index];

  // moving to same list
  if (source.droppableId === destination.droppableId) {
    const reordered: Quote[] = reorder(
      current,
      source.index,
      destination.index,
    );
    const result: QuoteMap = {
      ...quoteMap,
      [source.droppableId]: reordered,
    };
    return {
      quoteMap: result,
    };
  }

  // moving to different list

  // remove from original
  current.splice(source.index, 1);
  // insert into next
  next.splice(destination.index, 0, target);

  const result: QuoteMap = {
    ...quoteMap,
    [source.droppableId]: current,
    [destination.droppableId]: next,
  };

  return {
    quoteMap: result,
  };
};

