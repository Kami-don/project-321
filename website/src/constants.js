// @flow
export const colors = {
  // background
  dark500: '#151625',
  // foreground
  dark400: '#323543',
  // no defined role (purple grey)
  dark300: '#6F6F91',
  // subheading
  dark200: '#9D9DB7',
  // headings, text color
  dark100: '#D6DBF3',

  // blues
  blue500: '#004178',
  blue400: '#1573C8',
  blue300: '#29BCF3',

  // greens
  green500: '#033F55',
  green400: '#00C584',
  green300: '#00FFAD',

  // purples / reds
  purple500: '#4C0D31',
  purple400: '#982957',
  purple300: '#723874',
};

export const borderRadius: number = 2;
export const sidebarWidth: number = 250;
export const contentWidth: number = 960;
export const lineHeight: number = 1.5;
export const fontSizeEm: number = 1.2;
export const grid: number = 8;
export const gutter = {
  normal: grid * 2,
  large: grid * 4,
};
