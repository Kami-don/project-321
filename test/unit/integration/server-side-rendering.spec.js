/**
* @jest-environment node
*/
// @flow
import React, { Component } from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import { DragDropContext, Droppable, Draggable } from '../../../src/';
import type { Provided as DroppableProvided } from '../../../src/view/droppable/droppable-types';
import type { Provided as DraggableProvided } from '../../../src/view/draggable/draggable-types';

class App extends Component<*, *> {
  onDragStart = () => {
    // drag is starting!
  }

  onDragEnd = () => {
    // drag is ending!
  }

  render() {
    return (
      <DragDropContext
        onDragStart={this.onDragStart}
        onDragEnd={this.onDragEnd}
      >
        <Droppable droppableId="drop-1">
          {(provided: DroppableProvided) => (
            <div ref={provided.innerRef} {...provided.droppableProps}>
              <Draggable draggableId="drag-1">
                {(dragProvided: DraggableProvided) => (
                  <div>
                    <div
                      ref={dragProvided.innerRef}
                      {...dragProvided.draggableProps}
                      {...dragProvided.dragHandleProps}
                    />
                    {dragProvided.placeholder}
                  </div>
                )}
              </Draggable>
            </div>
          )}
        </Droppable>
      </DragDropContext>
    );
  }
}

describe('server side rendering', () => {
  // Checking that the browser globals are not available in this test file
  if (typeof window !== 'undefined' || typeof document !== 'undefined') {
    throw new Error('browser globals found in node test');
  }

  it('should support rendering to a string', () => {
    const result: string = renderToString(<App />);

    expect(result).toEqual(expect.any(String));
    expect(result).toMatchSnapshot();
  });

  it('should support rendering to static markup', () => {
    const result: string = renderToStaticMarkup(<App />);

    expect(result).toEqual(expect.any(String));
    expect(result).toMatchSnapshot();
  });
});
