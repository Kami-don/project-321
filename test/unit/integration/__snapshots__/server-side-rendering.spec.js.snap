// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server side rendering should support rendering to a string 1`] = `"<div data-react-beautiful-dnd-droppable=\\"0\\" data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"-1823773523\\"><div data-reactid=\\"2\\"><div data-react-beautiful-dnd-draggable=\\"0\\" tabindex=\\"0\\" data-react-beautiful-dnd-drag-handle=\\"0\\" aria-roledescription=\\"Draggable item. Press space bar to lift\\" draggable=\\"false\\" data-reactid=\\"3\\"></div></div></div>"`;

exports[`server side rendering should support rendering to static markup 1`] = `"<div data-react-beautiful-dnd-droppable=\\"1\\"><div><div data-react-beautiful-dnd-draggable=\\"1\\" tabindex=\\"0\\" data-react-beautiful-dnd-drag-handle=\\"1\\" aria-roledescription=\\"Draggable item. Press space bar to lift\\" draggable=\\"false\\"></div></div></div>"`;
